pep-0707.rst

```diff
@@ -0,0 +1,358 @@
+a safe, gradual and minimally disruptive way to make this change in the
+case of ``__exit__``.
+
+It will take multiple releases to get to a point where we can think of
+deprecating ``sys.exc_info()``. However, we can relatively quickly reach a
+stage where new learners do not need to know about it, or about the 3-tuple
+representation, at least until they are maintaining legacy code.
+
+
+Specification
+=============
+
+A context manager's ``__exit__``/``__aexit__`` method can have a single-arg
+signature, in which case it is invoked by the interpreter with the argument
+equal to an exception instance or ``None``:
+
+.. code-block::
+
+   >>> class C:
+   ...     def __enter__(self):
+   ...         return self
+   ...     def __exit__(self, exc):
+   ...         print('__exit__ called with:', repr(exc))
+   ... 
+   >>> with C():
+   ...     pass
+   ... 
+   __exit__ called with: None
+   >>> with C():
+   ...     1/0
+   ... 
+   __exit__ called with: ZeroDivisionError('division by zero')
+   Traceback (most recent call last):
+     File "<stdin>", line 2, in <module>
+   ZeroDivisionError: division by zero
+   >>> 
+
+
+If ``__exit__``/``__aexit__`` has any other signature, it is invoked with
+the 3-tuple ``(typ, exc, tb)`` as happens now:
+
+.. code-block::
+
+   >>> class C:
+   ...     def __enter__(self):
+   ...         return self
+   ...     def __exit__(self, *exc):
+   ...         print('__exit__ called with:', repr(exc))
+   ... 
+   >>> with C():
+   ...     pass
+   ... 
+   __exit__ called with: (None, None, None)
+   >>> with C():
+   ...     1/0
+   ... 
+   __exit__ called with: (<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero'), <traceback object at 0x1039cb570>)
+   Traceback (most recent call last):
+     File "<stdin>", line 2, in <module>
+   ZeroDivisionError: division by zero
+
+
+These ``__exit__`` methods will also be called with a 3-tuple:
+
+.. code-block::
+
+   ...     def __exit__(self, typ, *exc):
+   ...         print('__exit__ called with:', repr(typ, exc))
+   ...
+   ...     def __exit__(self, typ, exc, tb):
+   ...         print('__exit__ called with:', repr(type, exc, tb))
+
+
+A reference implementation is provided at [https://github.com/python/cpython/pull/101995].
+
+When the interpreter reaches the end of the scope of a context manager,
+and it is about to call the relevant ``__exit__`` or ``__aexit__`` function,
+it instrospects this function to determine whether it is the single-arg
+or the legacy 3-arg version. In the draft PR, this introspection is performed
+by ``is_legacy___exit__``, whose code is included below.
+
+It is important to note that this is not a generic introspection function, but
+rather one which is specifically designed for our use case. We know that
+``exit_func`` is an attribute of the context manager class (taken from the
+type of the object that provided ``__enter__``), so it is an unbound method.
```

Comment:
Or some other callable (perhaps created by a decorator).

Reply:
I'll just say function (it's not any callable).

Reply:
Non-function callables work:

```
In [1]: class Exiter:
   ...:     def __call__(self, *args):
   ...:         print(args)
   ...: 

In [2]: class X:
   ...:     def __enter__(self): return None
   ...:     __exit__ = Exiter()
   ...: 

In [4]: with X(): pass
(None, None, None)
```

Sorry for the nitpicky comment, I just have a feeling that silly edge cases like this could endanger the proposal.

Reply:
Ok I see what you mean.

> I just have a feeling that silly edge cases like this could endanger the proposal.

AFAIK perfect introspection is not possible, so there will be such edge cases (even if we make this one work). But the point is that they are limitations on new code, not something that breaks old code. So I don't know if they necessarily kill the proposal.